받은 후기 목록 조회

sujeong11 opened the pull request:
## 관련 이슈
- (관련 Issue를 닫아야 하는 경우) Close #이슈번호
- (관련 Issue를 닫을 필요가 없는 경우) #이슈번호

Close #42

## 작업 사항
- 작업사항 작성

  - **로그인한 사용자의 아이디(PK) 값으로 받은 후기 목록 조회하는 기능을 구현했습니다.**

## 참고 자료
- 참고자료 (링크 등)

## 체크리스트
- [X]  PR의 제목은 팀 내 규칙을 준수하여 알맞게 작성하였는가?
- [X]  Merge 하는 브랜치가 올바른가? (`main` branch에 실수로 PR 생성 금지)
- [X]  팀의 코딩 컨벤션을 준수하는가?
- [X]  PR과 관련없는 변경사항이 들어가지는 않았는가?
- [X]  내 코드에 대한 자기 검토가 되었는가?
- [X]  Reviewers, Assignees, Lables, Project, Milestone은 적절하게 선택하였는가?
- [X]  관련한 issue를 닫아야 하는지 점검해보고 적용했는가?

sujeong11 committed:
feat: #42 받은 후기 목록 조회 API
 리뷰 테이블 receiver의 id(PK)로 조회하는 쿼리 메서드에 로그인된 사용자의 id를 값을 주어 받은  후기 목록 전체를 조회해온다. 그리고 나서 프론트분들에게 전달해야 하는 데이터들을 가진 리스트타입의 DTO에 앞서 조회해온 목록들을 돌며 값을 넣어 반환한다.

sujeong11 committed:
feat: #42 swagger 문서화 추가

sujeong11 committed:
style: #42 오타 수정 및 불필요한 코드 삭제

Wo-ogie commented on src/main/java/com/umc/withme/dto/review/ReviewGetInfoResponse.java:
```diff
@@ -0,0 +1,29 @@
+
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@Getter
+public class ReviewGetInfoResponse {
```
`Get`이라는 이름이 꼭 들어가지 않아도 괜찮다면 `ReviewInfoResponse` 정도는 어떨까요? 
모임쪽 API를 구현하다보니 `MemberGetInfoResponse`를 다른 응답에서 사용해야하는 경우가 있었는데, 그러다보니 `Get`이라는 class name이 조금 거슬리더라구요. 저는 `Get`이라는 이름을 빼더라도 이름이 갖는 본연의 역할을 다 하는 것 같아서 그렇게 할까 고민중이긴 한데, 어떻게 생각하세요?

sujeong11 replied:
Get을 빼도 이해하는데 어려움이 없을 것 같습니다!

Wo-ogie replied:
감사합니다! 그럼 혹시 제가 작업하면서 `MemberGetInfoResponse`도 `MemberInfoResponse`로 수정해도 괜찮을까요?

sujeong11 replied:
넵!

Wo-ogie commented on src/main/java/com/umc/withme/controller/ReviewController.java:
```diff
@@ -48,4 +50,20 @@ public ResponseEntity<DataResponse<ReviewCreateResponse>> createReview(
                 .body(new DataResponse<>(response));
     }
+
+    @Operation(
+            summary = "받은 후기 글 조회",
+            description = "로그인된 사용자의 아이디(pk)로 받은 후기를 조회합니다.",
+            security = @SecurityRequirement(name = "access-token")
+    )
+    @GetMapping("/reviews/send-reviews")
+    public ResponseEntity<DataResponse<List<ReviewGetInfoResponse>>> getReceiveReview(
```
API 설명과 URL이 서로 매칭이 안되는 것 같은데, 맞을까요?
받은 리뷰인데, url이 `send-reviews`로 되어 있어서요.

dldmldlsy replied:
메소드명 getReceivedReviews는 바꾸는 건 어떨까요??

Wo-ogie commented on src/main/java/com/umc/withme/service/ReviewService.java:
```diff
@@ -49,4 +53,23 @@ public Long create(Long senderId, Long receiverId, Long meetId, ReviewCreateRequ
+     * @return ReviewGetInfoResponse 타입의 리스트
+     */
+    public List<ReviewGetInfoResponse> getReceiveReview(Long id) {
+        List<ReviewGetInfoResponse> infoList = new ArrayList<>();
+
+        reviewRepository.findAllByReceiver_Id(id)
+                .stream()
+                .forEach((el) -> infoList.add(ReviewGetInfoResponse.of(el.getSender().getNickname(),
+                                                                        el.getContent(),
+                                                                        el.getMeet().getTitle(),
+                                                                        el.getCreatedAt())));
+
+        return infoList;
```
`ReviewDto`가 있으니, `ReviewGetInfoResponse` 클래스에 `ReviewDto`를 parameter로 전달받아 응답 객체를 생성하는 `from()` factory method만 정의해준다면, `map()`을 활용해 다음과 같이 코드를 작성하는 것도 가능할 것 같습니다.

```java
return reviewRepository.findAllByReceiver_Id(id).stream()
                .map(ReviewDto::from)  // 조회된 Review entity를 전부 ReviewDto로 변경
                .map(ReviewGetInfoResponse::from)  // 앞서 ReviewDto로 변경한 모든 객체를 ReviewGetInfoResponse로 변경
                .collect(Collectors.toUnmodifiableList());  // mapping 결과를 list로 변환
```

sujeong11 replied:
우기님! 제시해주신 리뷰대로 로직 수정해보려고 했습니다 그런데 DTO로 바꿀 때 자꾸 NullPointer 에러가 발생하는데 아무리 봐도 해결이 안되더라고요.. 에러난 코드 일단 commit해두었고 시간나시면 봐주시면 감사하겠습니당😀

Wo-ogie replied:
에러 나는 부분이 그 부분만 있는걸까요?
코드 확인해보니, DTO내에 `from` method에서 모임 정보를 생성할 때 주소랑 모임 리더를 `null`로 넣고 계시더라구요. `MeetDto.from()` method 보시면 아시겠지만 주소와 리더에 대해 여러 method를 사용하고 있습니다. 이 중 객체 또는 필드가 `null`인 경우 에러가 발생할 수 있는 로직이 다수 있구요. 기본적으로 `null`이 아니라는 가정하에 짜여진 코드고, null-checking은 안된 상태입니다.

일단 해당 DTO의 로직을 변경하지 않는다는 가정하에 방법은 세 가지 정도가 있을 것 같습니다.
1. `Meet`의 스펙을 변경. `Address`와 양방향 연결을 하고, 필드에 주소 목록과 모임 리더를 추가하는 방법. 이렇게 하면 다음과 같이 DTO 변환이 가능해집니다.
```java
public static ReviewDto from(Review review) {
    return new ReviewDto(
            review.getId(),
            MemberDto.from(review.getSender()),
            MemberDto.from(review.getReceiver()),
            MeetDto.from(
                    review.getMeet(),
                    review.getMeet().getAddresses(),
                    review.getMeet().getLeader()
            ),
            PointDto.from(review.getPoint()),
            review.getContent(),
            review.getCreatedAt()
    );
}
```
2. `from` method의 parameter로 필요한 데이터를 받는 방법. 이 경우 DTO로 변환하기 전에 `Reivew` entity가 갖고 있는 `Meet` entity를 활용하여 모임의 주소 목록과 모임 리더를 조회해야겠죠?
```java
public static ReviewDto from(
        Review review,
        List<Address> addresses,
        Member leader
) {
    return new ReviewDto(
            review.getId(),
            MemberDto.from(review.getSender()),
            MemberDto.from(review.getReceiver()),
            MeetDto.from(review.getMeet(), addresses, leader),
            PointDto.from(review.getPoint()),
            review.getContent(),
            review.getCreatedAt()
    );
}
```
3.  모임 자체에 `null`을 넣어버리는 방법. 당장에 `ReviewInfoResponse`만을 생성하기 위해서라면 이렇게 해도 문제는 없습니다. 다만, 장기적인 관점에서 봤을 때는 DTO든 Entity든 구조적인 고민이 필요해 보여요.
```java
public static ReviewDto fromWithNoMeet(
        Review review
) {
    return new ReviewDto(
            review.getId(),
            MemberDto.from(review.getSender()),
            MemberDto.from(review.getReceiver()),
            null,
            PointDto.from(review.getPoint()),
            review.getContent(),
            review.getCreatedAt()
    );
}
```
4. `MeetDto`내에서 주소와 리더를 `null`로 넣는 것을 허용하는 `from()` method를 새로 정의하기. 이건 3번과 비슷한 내용입니다. 데이터에 `null`을 넣는 방식이죠? 물론 기존 코드와의 호환성도 고려해야 하고, 기존에 사용하던 `from` method도 필요한 method이니 method overloading 하는 방식으로 적용하면 좋을 것 같습니다.

일단... 모임 DTO의 로직을 변경하지 않는다는 가정하에  이렇게 네 가지 정도를 고려해 볼 수 있을 것 같구요. 저도 시간이 많이는 없어서 "뭐가 좋을 것 같다" 딱 말할 정도로 고민을 해보지는 않았어요 😅. 물론 위 방법들 말고도 여러 가지 방법들이 더 있을 것 같기는 한데, 일단 당장 급하게 해결하는 방법으로는 저 정도가 생각나네요. 파크가 한 번 고민해보시고, 좋은 코드로 작성해주시면 좋을 것 같습니다!

sujeong11 replied:
다양한 해결법이 있었네요! 알려주셔서 진짜 감사해요ㅠ😀 일단 4번 방법이 제일 간단하고 직관적이라 이 방법으로 작성해두었습니다ㅎㅎ

Wo-ogie commented on src/main/java/com/umc/withme/service/ReviewService.java:
```diff
@@ -49,4 +53,23 @@ public Long create(Long senderId, Long receiverId, Long meetId, ReviewCreateRequ
+     * @param id 로그인한 사용자의 아이디
+     * @return ReviewGetInfoResponse 타입의 리스트
+     */
+    public List<ReviewGetInfoResponse> getReceiveReview(Long id) {
```
약간의 의미차이이긴 한데, 응답 데이터가 리스트이고 받은 리뷰니까 method name으로 `getReceivedReviews`는 어떨까요?

Wo-ogie replied:
그리고, service layer의 method에서 id(PK)를 받을때는 `loginMemberId`, `receiverId`, `reviewId`, `meetId` 등 어떤 도메인의 PK 값인지 의식적으로 명시해주는게 나중에 코드 가독성이 좋더라구요! 아무래도 `id`가 공통적으로 사용되는 naming이다보니 참고해주셔도 좋을 것 같습니다 😊

sujeong11 replied:
넵 참고해서 수정하겠습니다!

sujeong11 committed:
style: #42 코드리뷰 반영
- 파일명, 변수명 명시적으로 변경
- API 경로 올바르게 수정
- DTO 사용하도록 로직 변경: NullPointer 에러 아직 해결 못함

sujeong11 committed:
fix: #42 Dto 변환 에러
